nixos/modules/contracts/streamingBackup.nix

```diff
@@ -0,0 +1,184 @@
+            type = str;
+            default = "db.bck";
+          };
+        };
+
+        config = lib.mkMerge [
+          (lib.setAttrByPath providerRoot {
+            consumer = config.services.postgresql.streamingBackup;
+          })
+          {
+            services.postgresql = {
+              streamingBackup.provider = lib.getAttrFromPath providerRoot;
+
+              enable = true;
+              ensureDatabases = [
+                config.test.username
+                "testdb"
+              ];
+              ensureUsers = [{
+                name = config.test.username;
+                ensureDBOwnership = true;
+                ensureClauses.login = true;
+              }];
+            };
+            test.backupName = "db.bck";
+            test.username = config.services.postgresql.superUser;
+          }
+          (lib.mkIf (config.test.username != "root") {
+            users.users.${config.test.username} = {
+              isSystemUser = true;
+              group = config.test.username;
+            };
+            users.groups.${config.test.username} = {};
+          })
+        ];
+      };
+
+      testScript = { nodes, ... }: let
+        cfg = nodes.machine;
+        inherit (lib.getAttrFromPath providerRoot nodes.machine) output;
+      in ''
+        import csv
+
+        start_all()
+        machine.wait_for_unit("postgresql.service")
+        machine.wait_for_open_port(5432)
+
+        def peer_cmd(cmd, db="testdb"):
+            return "sudo -u ${cfg.test.username} psql -U ${cfg.test.username} {db} --csv --command \"{cmd}\"".format(cmd=cmd, db=db)
+
+        def query(query):
+            res = machine.succeed(peer_cmd(query))
+            return list(dict(l) for l in csv.DictReader(res.splitlines()))
+
+        def cmp_tables(a, b):
+            for i in range(max(len(a), len(b))):
+                diff = set(a[i]) ^ set(b[i])
+                if len(diff) > 0:
+                    raise Exception(i, diff)
+
+        table = [{'name': 'car', 'count': '1'}, {'name': 'lollipop', 'count': '2'}]
+
+        with subtest("create fixture"):
+            machine.succeed(peer_cmd("CREATE TABLE test (name text, count int)"))
+            machine.succeed(peer_cmd("INSERT INTO test VALUES ('car', 1), ('lollipop', 2)"))
+
+            res = query("SELECT * FROM test")
+            cmp_tables(res, table)
+
+        with subtest("backup"):
+            print(machine.succeed("systemctl start ${output.backupService}"))
+
+        with subtest("drop database"):
+            print(machine.succeed(peer_cmd("DROP DATABASE testdb", db="postgres")))
+            machine.fail(peer_cmd("SELECT * FROM test"))
+
+        with subtest("restore"):
+            print(machine.succeed("${output.restoreScript} restore latest"))
+
+        with subtest("check restoration"):
+            res = query("SELECT * FROM test")
+            cmp_tables(res, table)
+      '';
+    };
+  };
+
+  meta.buildDocsInSandbox = false;
+}
```

Comment:
thought: Neither of the backup contracts covers any scheduling of backups or integration with file system snapshots. Maybe that's ok for now, or should be covered by something else anyway.

Reply:
Contracts are for communication between the consumer and provider, without the end user intervention. I see the file backup contract here only useful for the consumer to tell what files should be backed up or not. 

On the other hand, I see scheduling as the end user communicating with the provider directly. I could see a need for a common interface for scheduling though. So maybe a separate contract for that?

Now I could be wrong and there is a valid use case for the consumer having a say in the scheduling, I just did not cross it yet.